- fix building on FreeBSD

madame-rachelle opened the pull request:
I'm doing this via pull request mainly because it modifies a vital file which will cause many files of the project to be rebuilt once this is merged.

This fixes FreeBSD not being able to find vital locale-specific functions such as `tolower()` and `isdigit()` etc.

alexey-lysiuk commented on src/doomdef.h:
```diff
@@ -30,7 +30,9 @@
-
+#ifdef __FreeBSD__
+#include <ctype.h>
+#endif
```
I think `ctype.h` can be included unconditionally.

BTW, are those functions used so frequently? I mean how many .cpp files should be changed?

madame-rachelle replied:
They're used in at least two files, if not more. I don't know how many files they're actually used in, but with the localization stuff not even being finished yet, and those functions often being regarded as "convenience functions" - it would not surprise me if they pop up more, in the future. I honestly don't know of a better way to handle this situation - the only CI service that handles FreeBSD is Cirrus-CI, so if someone does another change again somewhere else, unless that's added to this repo it will just break again anyhow.

As for including it unconditionally - true that might be viable, but I chose not to do it because it wasn't necessary. Only one platform needs it right now, that I know of.

alexey-lysiuk replied:
It will be included on all platforms anyway, just indirectly.

madame-rachelle replied:
Which makes me even more leery of including it unconditionally. But that's just me. Until either you or Graf tell me directively to remove the #ifdef guard, I will leave it there, because like I said, I am not sure what's really the best way to handle it. I don't feel great about including a file if it's already included, and this is really only needed for FreeBSD right now.

alexey-lysiuk commented on src/gamedata/resourcefiles/file_wad.cpp:
```diff
@@ -38,7 +38,6 @@
 #include "w_wad.h"
 #include "gi.h"
 #include "doomerrors.h"
-
```
Was it an intentional change?

madame-rachelle replied:
No, that wasn't. The file was changed in the process of trying to fix the compile, but spaces are hard to remember when healing the file back after changing it.

coelckers commented:
I absolutely loathe conditional includes - or any kind of #ifdef boilerplate at the top of source files to make the code "platform independent". They cause nothing but problems and I consider this a poor way of doing stuff, even it it seems very popular in the Unix world, especially among old-school C (not ++) programmers. Code like this will always exhibit problems down the line. Just as a recent example, the xlsx reader I added a few days ago was a mess of platform dependent definitions where I had to fix a lot to make it compile on Windows - and I already fixed the biggest mess in it when I integrated it into an iOS project at work a year ago. That still didn't help making it compile on Windows.

On Windows ctype.h is part of the precompiled header anyway and if there's places where it works on Linux but not on FreeBSD, these places should be fixed explicitly to remove dependencies on indirect includes.

madame-rachelle commented:
> 
> 
> I absolutely loathe conditional includes - or any kind of #ifdef boilerplate at the top of source files to make the code "platform independent". They cause nothing but problems and I consider this a poor way of doing stuff, even it it seems very popular in the Unix world, especially among old-school C (not ++) programmers. Code like this will always exhibit problems down the line. Just as a recent example, the xlsx reader I added a few days ago was a mess of platform dependent definitions where I had to fix a lot to make it compile on Windows - and I already fixed the biggest mess in it when I integrated it into an iOS project at work a year ago. That still didn't help making it compile on Windows.
> 
> On Windows ctype.h is part of the precompiled header anyway and if there's places where it works on Linux but not on FreeBSD, these places should be fixed explicitly to remove dependencies on indirect includes.

So by this - are you saying `#include <ctype.h>` should simply be put specifically in the files where it is needed?

coelckers commented:
Yes, precisely that.

I really hope the C++ committee finally gets their act together and finalizes the module system so that this #include clusterfuck can be retired for good.

madame-rachelle commented:
> I really hope the C++ committee finally gets their act together and finalizes the module system so that this #include clusterfuck can be retired for good.

Doubt that will ever happen.

Nevertheless - this is changed now to fit the requests of both of you.
